src/common/messaging/gateway.ts

```diff
@@ -0,0 +1,342 @@
+	 */
+	public async fetch<T extends JsonValue = JsonValue>(path: string, body?: JsonValue): Promise<MessageResponse<T>>;
+	/**
+	 * Sends the {@link requestOrPath} to the server; the server should be listening on {@link MessageGateway.route}.
+	 * @param requestOrPath The request, or the path of the request.
+	 * @param bodyOrUndefined Request body, or moot when constructing the request with {@link MessageRequestOptions}.
+	 * @returns The response.
+	 */
+	public async fetch<T extends JsonValue = JsonValue>(requestOrPath: MessageRequestOptions | string, bodyOrUndefined?: JsonValue): Promise<MessageResponse<T>> {
+		const id = crypto.randomUUID();
+		const { body, path, timeout = DEFAULT_TIMEOUT, unidirectional = false } = typeof requestOrPath === "string" ? { body: bodyOrUndefined, path: requestOrPath } : requestOrPath;
+
+		// Initialize the response handler.
+		const response = new Promise<MessageResponse<T>>((resolve) => {
+			this.requests.set(id, (res: MessageResponse) => {
+				if (res.status !== 408) {
+					clearTimeout(timeoutMonitor);
+				}
+
+				resolve(res as MessageResponse<T>);
+			});
+		});
+
+		// Start the timeout, and send the request.
+		const timeoutMonitor = setTimeout(() => this.handleResponse({ __type: "response", id, path, status: 408 }), timeout);
+		const accepted = await this.proxy({ __type: "request", body, id, path, unidirectional } satisfies RawMessageRequest);
+
+		// When the server did not accept the request, return a 406.
+		if (!accepted) {
+			this.handleResponse({ __type: "response", id, path, status: 406 });
+		}
+
+		return response;
+	}
+
+	/**
+	 * Attempts to process the specified {@link message}.
+	 * @param message Message to process.
+	 * @returns `true` when the {@link message} was processed by this instance; otherwise `false`.
+	 */
+	public async process(message: DidReceivePluginMessage<JsonValue> | DidReceivePropertyInspectorMessage<JsonValue>): Promise<void> {
+		if (isRequest(message.payload)) {
+			// Server-side handling.
+			const action = this.actionProvider(message);
+			if (await this.handleRequest(action, message.payload)) {
+				return;
+			}
+
+			this.emit("unhandledRequest", message as DidReceivePluginMessage<RawMessageRequest> | DidReceivePropertyInspectorMessage<RawMessageRequest>);
+		} else if (isResponse(message.payload) && this.handleResponse(message.payload)) {
+			// Response handled successfully.
+			return;
+		}
+
+		this.emit("unhandledMessage", message);
+	}
+
+	/**
+	 * Maps the specified {@link path} to the {@link handler}, allowing for requests from the client.
+	 * @param path Path used to identify the route.
+	 * @param handler Handler to be invoked when the request is received.
+	 * @param options Optional routing configuration.
+	 * @returns This instance with the route registered.
+	 */
+	public route<TBody extends JsonValue = JsonValue>(path: string, handler: UnscopedMessageHandler<TAction, TBody>, options?: RouteConfiguration<TAction>): this {
```

Comment:
Is there ever any circumstance where we might want to allow a route to be unregistered?

Reply:
There is, but not at the moment. When transient actions are available, and they're added / removed from memory as needed, we'll need to ensure the routes for those context/action combinations are also removed appropriately. This also drove the decision to _not_ return `this` on the publicly exposed `registerRoute` as that will likely return a `Disposable` to allow for the route to be removed.
https://github.com/elgatosf/streamdeck/blob/d41df3a2e75f4dd0dd6843517a076dbd38547a2d/src/plugin/ui/controller.ts#L77-L83

Reply:
`registerRoute` now returns an `IDisposable` 🥳